crypto/pkcs11key/key.go

```diff
@@ -252,10 +312,117 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
 	if !gotModulus || !gotExponent {
 		return noKey, errors.New("public key missing either modulus or exponent")
 	}
-	return rsa.PublicKey{
+
+	rsa := rsa.PublicKey{
 		N: n,
 		E: e,
-	}, nil
+	}
+	return rsa, nil
+}
+
+// Get the public key matching an Elliptic Curve private key
+func (ps *Key) getECPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkcs11.ObjectHandle) (interface{}, error) {
+	var noKey interface{}
+
+	template := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_EC_PARAMS, nil),
+		pkcs11.NewAttribute(pkcs11.CKA_ID, nil),
+	}
+
+	attr, err := module.GetAttributeValue(session, privateKeyHandle, template)
+	if err != nil {
+		return noKey, err
+	}
+
+	oid := []byte {}
+	id := []byte {}
+	gotOid, gotID := false, false
+	for _, a := range attr {
+		if a.Type == pkcs11.CKA_EC_PARAMS {
+			oid = a.Value
+			gotOid = true
+		} else if a.Type == pkcs11.CKA_ID {
+			id = a.Value
+			gotID = true
+		}
+	}
+	if !gotOid || !gotID {
+		return noKey, errors.New("public key missing either curve parameters or id")
+	}
+
+	poid := new(asn1.ObjectIdentifier)
+	asn1.Unmarshal(oid, poid)
+	curve := namedCurveFromOID(*poid)
+
+	templateSearch := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PUBLIC_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_ID, id),
+	}
+
+	if err := module.FindObjectsInit(session, templateSearch); err != nil {
+		return noKey, err
+	}
+	objs, _, err := module.FindObjects(session, 2)
+	if err != nil {
+		return noKey, err
+	}
+	if err = module.FindObjectsFinal(session); err != nil {
+		return noKey, err
+	}
+
+	if len(objs) == 0 {
+		return noKey, fmt.Errorf("public key not found")
+	}
+	publicKeyHandle := objs[0]
+
+	templatePub := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_EC_POINT, nil),
+	}
+
+	attrPub, err := module.GetAttributeValue(session, publicKeyHandle, templatePub)
+	if err != nil {
+		return noKey, err
+	}
+
+	data := []byte {}
+	gotData := false
+	for _, a := range attrPub {
+		if a.Type == pkcs11.CKA_EC_POINT {
+			data = a.Value
+			gotData = true
+		}
+	}
+	if !gotData {
+		return noKey, errors.New("public key missing EC Point")
+	}
+
+	x, y := elliptic.Unmarshal(curve, data)
+	if x == nil && y == nil {
```

Comment:
Per docs for `elliptic.Unmarshal`: "On error, x = nil". It's confusing to check for both; better to just check for the one that indicates the error condition.

Can you also document the fallback parsing that is done below? When do you expect it to trigger, and what format does it expect? Would be good to break out this part of the function (from line 399 to line 407) into its own function for easier testing.